src/pdn/src/straps.cpp

```diff
@@ -413,6 +414,70 @@ FollowPins::FollowPins(Grid* grid, odb::dbTechLayer* layer, int width)
   }
 }
 
+std::unordered_map<odb::dbRow*, RowEndInsts> FollowPins::findRowEndInsts()
+{
+  std::unordered_map<odb::dbRow*, RowEndInsts> result;
+  auto rows = getDomain()->getRows();
+
+  std::unordered_map<int, odb::dbRow*> y_to_row;
+  for (auto row : rows) {
+    y_to_row[row->getBBox().yMin()] = row;
+  }
+
+  for (auto inst : getGrid()->getBlock()->getInsts()) {
+    if (!inst->isFixed()
+        || (!inst->isEndCap()
+            && inst->getMaster()->getType()
+                   != odb::dbMasterType::CORE_WELLTAP)) {
+      continue;
+    }
+    auto box = inst->getBBox();
+    auto found_row = y_to_row.find(box->yMin());
+    if (found_row == y_to_row.end()) {
+      continue;
+    }
+    odb::dbRow* row = found_row->second;
+    if (box->xMin() == row->getBBox().xMin()) {
+      result[row].left = inst;
+    }
+    if (box->xMax() == row->getBBox().xMax()) {
+      result[row].right = inst;
+    }
+  }
+
+  return result;
+}
+
+int FollowPins::getPinExtent(odb::dbInst* inst, odb::dbNet* net, bool right)
+{
+  // Find the lowest (right=false) or highest (right=true) x-coordinate of power
+  // pins in a cell
+  int extent = right ? inst->getBBox()->xMin() : inst->getBBox()->xMax();
+  const odb::dbTransform transform = inst->getTransform();
+
+  for (odb::dbITerm* iterm : inst->getITerms()) {
+    if (iterm->getNet() != net) {
+      continue;
+    }
+    for (odb::dbMPin* mpin : iterm->getMTerm()->getMPins()) {
+      for (odb::dbBox* box : mpin->getGeometry()) {
+        if (box->getTechLayer() != getLayer()) {
+          continue;
+        }
+        odb::Rect rect = box->getBox();
+        transform.apply(rect);
+        if (right) {
+          extent = std::max(extent, rect.xMax());
+        } else {
+          extent = std::min(extent, rect.xMin());
+        }
+      }
+    }
+  }
+
+  return extent;
+}
```

Comment:
it seems like what you really want is for each "row" to have a new BBox that is bounded by the pin on that layer (or by the row itself if no pins are available)
You may need to check that these pins actually overlap with the follow pin you are aiming for to avoid issues with m2 followpins on powerswitches.

Reply:
Do you have a good example of the m2 followpins case that I could use as a reference?

Reply:
For now, I've added a check to make sure the y-position overlaps.